### PR TITLE
Lower alloc KeyValueAccumulator

### PR DESCRIPTION
**Pull request by benaadams**
Avoid allocating `List<string>` per value 

@rynowak did you mean something like this in #553 ?

### CODE REVIEW

**Comment by rynowak** [comment]
Yeah, this is exactly suggesting.

One possible pivot here (and I don't think we need to do it without evidence that it's good) would be to use geometric growth on the backing `string[]` for `StringValues` and teach it to ignore null elements - but I'm not sure we need that.

File: src/Microsoft.AspNetCore.WebUtilities/KeyValueAccumulator.cs [hunk 1]
```diff
@@ -9,24 +9,33 @@ namespace Microsoft.AspNetCore.WebUtilities
...
             {
-                values.Add(value);
+                if (values.Count <= 1)
+                {
+                    _accumulator[key] = new string[] { values[0], value };
+                }
+                else
+                {
+                    var currentValues = values.ToArray();
+                    Array.Resize(ref currentValues, values.Count + 1);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.WebUtilities/KeyValueAccumulator.cs]: Yeah, this growth by one is what the List was intended to avoid.

Comment (benaadams) [inline src/Microsoft.AspNetCore.WebUtilities/KeyValueAccumulator.cs]: Will save allocs for values with counts  > 3 (or > 5 if you include lists internals); any idea on how often that occurs?
e.g. when I get into that situation I'll generally move to a json post as the array post is easier to reason about what's going on.

Comment (benaadams) [inline src/Microsoft.AspNetCore.WebUtilities/KeyValueAccumulator.cs]: There is the checkbox hack double pattern 
```html
<input name="foo" type="checkbox" value="true" />
<input name="foo" type="hidden" value="false" />
```
However this change should still have less allocations than `List` in this case

Comment (davidfowl) [inline src/Microsoft.AspNetCore.WebUtilities/KeyValueAccumulator.cs]: Yes this change isn't great. Easy to create a form that causes tons of resize calls.

Comment (benaadams) [inline src/Microsoft.AspNetCore.WebUtilities/KeyValueAccumulator.cs]: Betterz?

File: src/Microsoft.AspNetCore.WebUtilities/KeyValueAccumulator.cs [hunk 2]
```diff
@@ -9,24 +9,55 @@ namespace Microsoft.AspNetCore.WebUtilities
...
+                }
+                else
+                {
+                    // Third value for this key
+                    // Add zero count entry and move to data to expanding list dictionary
+                    _accumulator[key] = default(StringValues);
+
+                    if (_expandingAccumulator == null)
+                    {
+                        _expandingAccumulator = new Dictionary<string, List<string>>();
```
Comment (benaadams) [inline src/Microsoft.AspNetCore.WebUtilities/KeyValueAccumulator.cs]: needs `StringComparer.OrdinalIgnoreCase`

**Comment by benaadams** [comment]
Addressed multi-value growth issues.

**Comment by Tratcher** [comment]
@rynowak for verification.

**Comment by rynowak** [comment]
Sure bro

**Comment by rynowak** [comment]
This is definitely having the intended effect.

**Before**
image

**After**
image

Saves 19mb in this case.

**Comment by muratg** [comment]
@Tratcher Are we taking this?

**Comment by rynowak** [comment]
I'm not @Tratcher - but this is a huge perf win for big and small inputs

**Comment by davidfowl** [comment]
@Tratcher is this on your radar?

**Comment by Tratcher** [comment]
Yeah, it's next on the list...

**Comment by Tratcher** [comment]
Rebased, squashed, and merged.

**Pull request closed without merging**